src/wp-includes/default-filters.php

```diff
@@ -748,4 +748,9 @@
 // Font management.
 add_action( 'wp_head', 'wp_print_font_faces', 50 );
 
+// Modules.
+add_action( 'wp_head', 'wp_print_import_map' );
+add_action( 'wp_head', 'wp_print_enqueued_modules' );
+add_action( 'wp_head', 'wp_print_module_preloads' );
```

Comment:
Note: we need to move these prints to the footer on classic themes ([Gutenberg PR](https://github.com/WordPress/gutenberg/pull/57396)).

Reply:
I didn't like much [the solution used in Gutenberg](https://github.com/WordPress/gutenberg/pull/57396) because there's no need to print all the modules (or preloads) at once. So taking into account that the order in which modules are printed is not important, and that the sooner they are on the page the better because the browser will start downloading them sooner:

- I've modified the print functions which now keep track of the printed modules and preloads and therefore they can be called multiple times.
- I've added hooks to print both the enqueued and preloaded modules at both the `wp_head` and `wp_footer`. They will print the modules that are already available in the `wp_head` (like modules of blocks in block themes, or modules from classic themes or plugins) and they will print again the modules that were not available in the `wp_head` but are available in the `wp_footer` (like modules of blocks in classic themes).
- I've moved the import map to the footer to make sure that it includes all the dependencies.

This is the explanation included in the code:

```php
/**
 * Adds the hooks to print the import map, enqueued modules and module
 * preloads.
 *
 * It adds the actions to print the enqueued modules and module preloads to
 * both `wp_head` and `wp_footer` because in classic themes, the modules
 * used by the theme and plugins will likely be able to be printed in the
 * `head`, but the ones used by the blocks will need to be enqueued in the
 * `footer`.
 *
 * As all modules are deferred and dependencies are handled by the browser,
 * the order of the modules is not important, but it's still better to print
 * the ones that are available when the `wp_head` is rendered, so the browser
 * starts downloading those as soon as possible.
 *
 * The import map is also printed in the footer to be able to include the
 * dependencies of all the modules, including the ones printed in the footer.
 *
```

Reply:
@c4rl0sbr4v0 discovered that we can't enqueue or preload any module before we know all the dependencies because the import map needs to be added before any module:

- https://github.com/WordPress/gutenberg/pull/57778#discussion_r1450185051

To be honest, this is quite a surprise, especially the preload part. If someone knows better, please tell so.

I've already opened a Trac ticket to fix it: https://core.trac.wordpress.org/ticket/60240